filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -0,0 +1,272 @@
+	}
+
+	args = handleSeekAndCursor(args, mode, since, cursor)
+
+	for _, u := range units {
+		args = append(args, "--unit", u)
+	}
+
+	for _, i := range syslogIdentifiers {
+		args = append(args, "--identifier", i)
+	}
+
+	for _, m := range matchers.Matches {
+		args = append(args, m.String())
+	}
+
+	for _, m := range transports {
+		args = append(args, fmt.Sprintf("_TRANSPORT=%s", m))
+	}
+
+	logger.Debugf("Journalctl command: journalctl %s", strings.Join(args, " "))
+	cmd := exec.Command("journalctl", args...)
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return &Reader{}, fmt.Errorf("cannot get stderr pipe: %w", err)
+	}
+
+	r := Reader{
+		cmd:      cmd,
+		dataChan: make(chan []byte),
+		errChan:  make(chan error),
+		logger:   logger,
+		stdout:   stdout,
+		stderr:   stderr,
+		canceler: canceler,
+	}
+
+	// Goroutine to read errors from stderr
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stderr goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stderr)
+		msgs := []string{}
+		for {
+			line, err := reader.ReadString('\n')
+			if errors.Is(err, io.EOF) {
+				if len(msgs) == 0 {
+					return
+				}
+				errMsg := fmt.Sprintf("Journalctl wrote errors: %s", strings.Join(msgs, "\n"))
+				logger.Errorf(errMsg)
+				r.errChan <- errors.New(errMsg)
+				return
+			}
+			msgs = append(msgs, line)
+		}
+	}()
+
+	// Goroutine to read events from stdout
+	r.wg.Add(1)
+	go func() {
+		defer r.logger.Debug("stdout goroutine done")
+		defer r.wg.Done()
+		reader := bufio.NewReader(r.stdout)
+		for {
+			data, err := reader.ReadBytes('\n')
+			if errors.Is(err, io.EOF) {
+				close(r.dataChan)
+				return
+			}
+
+			select {
+			case <-r.canceler.Done():
+				return
+			case r.dataChan <- data:
+			}
+		}
+	}()
+
+	if err := cmd.Start(); err != nil {
+		return &Reader{}, fmt.Errorf("cannot start journalctl: %w", err)
+	}
+
+	return &r, nil
+}
+
+// Close stops the `journalctl` process and waits for all
+// goroutines to return, the canceller passed to `New` should
+// be cancelled before `Close` is called
+func (r *Reader) Close() error {
+	if r.cmd == nil {
+		return nil
+	}
+
+	if err := r.cmd.Process.Kill(); err != nil {
+		return fmt.Errorf("cannot stop journalctl: %w", err)
+	}
+
+	r.logger.Debug("waiting for all goroutines to finish")
+	r.wg.Wait()
+	return nil
+}
+
+// Next returns the next journal entry. If there is no entry available
+// next will block until there is an entry or cancel is cancelled.
+//
+// If cancel is cancelled, Next returns a zero value JournalEntry
+// and ErrCancelled.
+func (r *Reader) Next(cancel input.Canceler) (JournalEntry, error) {
+	select {
+	case <-cancel.Done():
+		return JournalEntry{}, ErrCancelled
+	case d, open := <-r.dataChan:
+		if !open {
+			return JournalEntry{}, errors.New("data chan is closed")
+		}
+		fields := map[string]string{}
```

Comment:
Thanks for also spotting that @fearful-symmetry!

I was as confident as the authors of github.com/coreos/go-systemd/sdjournal, because they read the messages as `map[string]string`: https://github.com/coreos/go-systemd/blob/7d375ecc2b092916968b5601f74cca28a8de45dd/sdjournal/journal.go#L405-L410 and our whole journald input.

But a little while ago, thanks to `kdeconnectd` crashing on my machine I found out some fields are not string, at least for core dumps.
`COREDUMP_PROC_AUXV` is a slice of bytes.

Our whole code after the reader assumes `map[string]string`. I'm fixing it right now, hopefully it won't be too bad to fix.